### PR TITLE
Rewrite jest unit tests to TypeScript

### PR DESCRIPTION
**Pull request by tjzel**
## Summary

- Renamed all tests with their respective TypeScript counterparts.
- Fixed jestUtils to properly allow strict checking of AnimatedStyle.
- Added type declaration emitting to plugin.
- Moved building plugin files before running TypeScript on the project (otherwise TypeScript tests would fail during type-checking).
- Added test for jestUtils.
- Bumped `jest` version and added `@types/react-test-renderer`.

---
Added to `jest-setup.js`:
```diff
+    delete global.MessageChannel;
```
This line is coming from `react-16-node-hanging-test-fix`. It fixes an error of open handles in `@testing-library/react-native` function `render` and jest going 😵. [Reference](https://github.com/facebook/react/issues/20756).

## Test plan

`yarn && yarn jest --detectOpenHandles` ftw

### CODE REVIEW

File: plugin/package.json [hunk 1]
```diff
@@ -12,12 +12,13 @@
...
   "scripts": {
     "prepare": "yarn generate",
-    "generate": "yarn tsc && yarn bundle && yarn clean",
+    "generate": "yarn tsc && yarn bundle && yarn format && yarn clean",
     "clean": "rm -f lib/*",
     "bundle": "esbuild ./lib/plugin.js --bundle --external:@babel --external:convert-source-map --external:fs --external:'../../package.json' --log-level=warning --platform=node --sourcemap=linked --outfile=build/plugin.js",
     "format": "prettier --write --list-different **/*.ts",
     "watch": "tsc-watch --onSuccess \"yarn bundle\"",
-    "lint": "eslint --max-warnings=0 --ext .ts .",
+    "lint": "eslint --max-warnings=0 --ext .ts",
```
Comment (tomekzaw) [inline plugin/package.json:20]:
```suggestion
    "lint": "eslint --max-warnings=0 --ext .ts .",
```

File: src/reanimated2/jestUtils.ts [hunk 2]
```diff
@@ -5,7 +5,10 @@ declare global {
   namespace jest {
     interface Matchers<R> {
       toHaveAnimatedStyle(
-        style: Record<string, unknown>[] | Record<string, unknown>
+        style: Record<string, unknown>[] | Record<string, unknown>,
+        config?: {
+          shouldMatchAllProps?: boolean;
```
Comment (tomekzaw) [inline src/reanimated2/jestUtils.ts:10]: this looks like a breaking change for library users, we'll need to discuss it with the team

File: __tests__/InterpolateColor.test.tsx [hunk 3]
```diff
@@ -153,21 +153,28 @@ describe('colors interpolation', () => {
...
 
     fireEvent.press(button);
     jest.advanceTimersByTime(250);
 
     expect(view).toHaveAnimatedStyle(
-      { backgroundColor: 'rgba(71, 117, 73, 1)' },
-      true
+      {
+        backgroundColor: 'rgba(71, 117, 73, 1)',
+      },
```
Comment (tomekzaw) [inline __tests__/InterpolateColor.test.tsx:167]:
```suggestion
      { backgroundColor: 'rgba(71, 117, 73, 1)' },
```

File: __tests__/InterpolateColor.test.tsx [hunk 4]
```diff
@@ -153,21 +153,28 @@ describe('colors interpolation', () => {
...
     );
 
     jest.advanceTimersByTime(250);
 
     expect(view).toHaveAnimatedStyle(
-      { backgroundColor: 'rgba(96, 144, 32, 1)' },
-      true
+      {
+        backgroundColor: 'rgba(96, 144, 32, 1)',
+      },
```
Comment (tomekzaw) [inline __tests__/InterpolateColor.test.tsx:176]:
```suggestion
      { backgroundColor: 'rgba(96, 144, 32, 1)' },
```

File: __tests__/jestUtils.test.tsx [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    expect(view).not.toHaveAnimatedStyle(
+      { flex: 1 },
+      { shouldMatchAllProps: true }
+    );
+
+    expect(view).toHaveAnimatedStyle(
+      { flex: 1, backgroundColor: 'black' },
+      { shouldMatchAllProps: false }
+    );
+
```
Comment (tomekzaw) [inline __tests__/jestUtils.test.tsx:41]:
```suggestion

    expect(view).toHaveAnimatedStyle(
      { flex: 1, backgroundColor: 'black' },
      { shouldMatchAllProps: true }
    );
```

File: package.json [hunk 6]
```diff
@@ -113,9 +113,10 @@
     "@types/babel__traverse": "^7.14.2",
     "@types/convert-source-map": "^2.0.0",
     "@types/invariant": "^2.2.35",
-    "@types/jest": "^27.4.0",
+    "@types/jest": "^29.0.0",
     "@types/node": "^18.0.0",
-    "@types/react": "^18.0.26",
+    "@types/react": "^18.0.0",
```
Comment (tomekzaw) [inline package.json:118]:
```suggestion
    "@types/react": "^18.0.26",
```

**Review by tomekzaw — CHANGES_REQUESTED** [review]

**Review by tomekzaw — APPROVED** [review]

**Comment by tjzel** [comment]
we wait with mergele

**Pull request merged**